### PR TITLE
[VectorCombine][X86] foldShuffleOfCastops - fold shuffle(cast(x),cast(y)) -> cast(shuffle(x,y)) iff cost efficient

### LINKED ISSUE #67803: [VectorCombine][X86] Poor handling of compare-select patterns with AVX2 spoofing on AVX1 targets
https://godbolt.org/z/Waonx44Mj

For AVX1 only targets we often encounter 'fake-AVX2' code for integer math like:
```c
#if !defined(__AVX2__)
#define _mm256_cmpgt_epi32( a, b ) \
 _mm256_setr_m128i( \
	_mm_cmpgt_epi32( _mm256_extractf128_si256( (a), 0 ), _mm256_extractf128_si256( (b), 0 ) ), \
	_mm_cmpgt_epi32( _mm256_extractf128_si256( (a), 1 ), _mm256_extractf128_si256( (b), 1 ) ) )

#define _mm256_blendv_epi8( a, b, c ) \
 _mm256_setr_m128i( \
	_mm_blendv_epi8( _mm256_extractf128_si256( (a), 0 ), _mm256_extractf128_si256( (b), 0 ), _mm256_extractf128_si256( (c), 0 ) ), \
	_mm_blendv_epi8( _mm256_extractf128_si256( (a), 1 ), _mm256_extractf128_si256( (b), 1 ), _mm256_extractf128_si256( (c), 1 ) ) )
#endif

__m256i cmpsel_epi8(__m256i x, __m256i y, __m256i a, __m256i b) {
    __m256i cmp = _mm256_cmpgt_epi32(x,y);
    return _mm256_blendv_epi8(a,b,cmp);
}
```
This is really poorly optimized, mainly due to all the bitcasts to/from the __m128i (<2 x i64>) types.

In particular we see this pattern a lot:
```ll
  %3 = bitcast <4 x i32> %sext.i to <2 x i64>
  %4 = bitcast <4 x i32> %sext.i21 to <2 x i64>
  %shuffle.i.i = shufflevector <2 x i64> %3, <2 x i64> %4, <4 x i32> <i32 0, i32 1, i32 2, i32 3>
  %7 = bitcast <4 x i64> %shuffle.i.i to <8 x i32>
```
We should be able to get VectorCombine to fold this to a <8 x i32> shufflevector instead, in fact VectorCombine::foldBitcastShuf might handle this if we extend it to binary shuffles, with improved cost handling.

We also see :
```ll
  %2 = icmp sgt <8 x i32> %0, %1
  %cmp.i = shufflevector <8 x i1> %2, <8 x i1> poison, <4 x i32> <i32 0, i32 1, i32 2, i32 3>
  %sext.i = sext <4 x i1> %cmp.i to <4 x i32>
  %3 = bitcast <4 x i32> %sext.i to <2 x i64>
  %cmp.i20 = shufflevector <8 x i1> %2, <8 x i1> poison, <4 x i32> <i32 4, i32 5, i32 6, i32 7>
  %sext.i21 = sext <4 x i1> %cmp.i20 to <4 x i32>
  %4 = bitcast <4 x i32> %sext.i21 to <2 x i64>
```
We've managed to combine to a single <8 x i32> icmp , but failed to rejoin the compare result sign extensions. We should be able to handle this in VectorCombine if we handle concatenation of casts (based off what we do in VectorCombine::foldShuffleOfBinops)

- [x] Extend VectorCombine::foldBitcastShuf to handle length changing shuffles
- [x] Extend VectorCombine::foldBitcastShuf to handle binary shuffles
- [ ] Add a VectorCombine::foldShuffleOfCasts similar to VectorCombine::foldShuffleOfBinops

### PR DESCRIPTION
**Pull request by RKSimon**
Based off the existing foldShuffleOfBinops fold

Fixes #67803

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/VectorCombine.cpp [hunk 1]
```diff
@@ -1432,6 +1433,59 @@ bool VectorCombine::foldShuffleOfBinops(Instruction &I) {
...
+                           m_Mask(Mask))))
+    return false;
+
+  auto *C0 = dyn_cast<CastInst>(V0);
+  auto *C1 = dyn_cast<CastInst>(V1);
+  if (!C0 || !C1)
+    return false;
+
+  Instruction::CastOps Opcode = C0->getOpcode();
+  if (Opcode == Instruction::BitCast || Opcode != C1->getOpcode() ||
```
Comment (goldsteinn) [inline llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1451]: elementwise bitcast should be okay?

File: llvm/lib/Transforms/Vectorize/VectorCombine.cpp [hunk 2]
```diff
@@ -1432,6 +1433,59 @@ bool VectorCombine::foldShuffleOfBinops(Instruction &I) {
...
+      C0->getSrcTy() != C1->getSrcTy())
+    return false;
+
+  auto *ShuffleDstTy = dyn_cast<FixedVectorType>(I.getType());
+  auto *CastDstTy = dyn_cast<FixedVectorType>(C0->getDestTy());
+  auto *CastSrcTy = dyn_cast<FixedVectorType>(C0->getSrcTy());
+  if (!ShuffleDstTy || !CastDstTy || !CastSrcTy ||
+      CastDstTy->getElementCount() != CastSrcTy->getElementCount())
```
Comment (goldsteinn) [inline llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1459]: Is this possible if you are skipping bitcasts?

File: llvm/lib/Transforms/Vectorize/VectorCombine.cpp [hunk 3]
```diff
@@ -1432,6 +1433,59 @@ bool VectorCombine::foldShuffleOfBinops(Instruction &I) {
...
+    return false;
+
+  auto *NewCastSrcTy =
```
Comment (goldsteinn) [inline llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1462]: nit: I read this more are `NewShuffleDstTy`

File: llvm/lib/Transforms/Vectorize/VectorCombine.cpp (hunk 1 above)
Comment (goldsteinn) [inline llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1451]: You will miss the `nneg zext` + `sext` case. Maybe add a todo?

Comment (RKSimon) [reply llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1451]: Eventually yes - except that we have other folds that fold bitcast(shuffle(x,y)) -> shuffle(bitcast(x),bitcast(y)) - I think what we're really after is whatever direction reduces the number of bitcasts, but we're not there yet :(

File: llvm/lib/Transforms/Vectorize/VectorCombine.cpp (hunk 2 above)
Comment (RKSimon) [reply llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1459]: I can replace it with an assertion.

File: llvm/lib/Transforms/Vectorize/VectorCombine.cpp [hunk 4]
```diff
@@ -1432,6 +1433,59 @@ bool VectorCombine::foldShuffleOfBinops(Instruction &I) {
...
+      Opcode, ShuffleDstTy, NewCastSrcTy, TTI::CastContextHint::None, CostKind);
+  NewCost += TTI.getShuffleCost(TargetTransformInfo::SK_PermuteTwoSrc,
+                                CastSrcTy, Mask, CostKind);
+  if (NewCost > OldCost)
+    return false;
+
+  Value *Shuf =
+      Builder.CreateShuffleVector(C0->getOperand(0), C1->getOperand(0), Mask);
+  Value *Cast = Builder.CreateCast(Opcode, Shuf, ShuffleDstTy);
+  // TODO Copy IR flags?
```
[the remark this answers is not in the document]
Comment (goldsteinn) [reply llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1488]: I think you should be able to do `copyIRFlags` + `andIRFlags`.

File: llvm/lib/Transforms/Vectorize/VectorCombine.cpp (hunk 1 above)
Comment (RKSimon) [reply llvm/lib/Transforms/Vectorize/VectorCombine.cpp]: I've added `nneg zext + sext` support

File: llvm/lib/Transforms/Vectorize/VectorCombine.cpp [hunk 5]
```diff
@@ -1432,6 +1433,79 @@ bool VectorCombine::foldShuffleOfBinops(Instruction &I) {
...
+    return false;
+
+  // Handle shuffle(zext_nneg(x), sext(y)) -> sext(shuffle(x,y)) folds.
+  if (Opcode != C1->getOpcode()) {
+    auto IsZExtNNeg = [](Instruction *I) {
+      auto *NNI = dyn_cast<PossiblyNonNegInst>(I);
+      return NNI && NNI->hasNonNeg();
+    };
+    if (!((C0->getOpcode() == Instruction::SExt && IsZExtNNeg(C1)) ||
+          (C1->getOpcode() == Instruction::SExt && IsZExtNNeg(C0))))
```
Comment (goldsteinn) [inline llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1461]: You can just use `m_SExtLike`

File: llvm/lib/Transforms/Vectorize/VectorCombine.cpp [hunk 6]
```diff
@@ -1432,6 +1433,75 @@ bool VectorCombine::foldShuffleOfBinops(Instruction &I) {
...
+      FixedVectorType::get(CastSrcTy->getScalarType(), Mask.size());
+
+  // Try to replace a castop with a shuffle if the shuffle is not costly.
+  TTI::TargetCostKind CostKind = TTI::TCK_RecipThroughput;
+
+  InstructionCost OldCost =
+      TTI.getCastInstrCost(C0->getOpcode(), CastDstTy, CastSrcTy,
+                           TTI::CastContextHint::None, CostKind) +
+      TTI.getCastInstrCost(C1->getOpcode(), CastDstTy, CastSrcTy,
+                           TTI::CastContextHint::None, CostKind);
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1480]: Can we avoid computing the cost twice? Is it needed by `zext nneg + sext`?
```suggestion
      TTI.getCastInstrCost(Opcode, CastDstTy, CastSrcTy,
                           TTI::CastContextHint::None, CostKind) * 2;
```

**Review by dtcxzyw — APPROVED** [review]
LGTM.

File: llvm/lib/Transforms/Vectorize/VectorCombine.cpp (hunk 6 above)
Comment (RKSimon) [reply llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1480]: Yes, zext can be notably cheaper than sext (that's why I had to tweak d53b8291bff4542a9c3e2f1df050deafbe295fff to more clearly show the cost diff on SSE vs AVX)

**Pull request merged**